python/ray/experimental/workflow/tests/test_virtual_actor_2.py

```diff
@@ -89,6 +89,214 @@ def test_indirect_actor_writer(workflow_start_regular):
+    lock.acquire()
+    fail_flag.unlink()
+    ret = actor.session_start_with_status.run_async()
+    for i in range(0, 60):
+        if cnt.read_text() == "4":
+            break
+        time.sleep(1)
+    assert cnt.read_text() == "4"
+    # This means when return from session_start_with_status,
+    # the session got updated
+    assert actor.get_status.run() == "STARTING"
+    lock.release()
+    assert ray.get(ret) == "UP"
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular",
+    [{
+        "num_cpus": 4
+        # We need more CPUs, otherwise 'create()' blocks 'get()'
+    }],
+    indirect=True)
+@pytest.mark.repeat(5)
+def test_wf_in_actor_chain(workflow_start_regular, tmp_path):
+    file_lock = [str(tmp_path / str(i)) for i in range(5)]
+    fail_flag = tmp_path / "fail"
+
+    @workflow.virtual_actor
+    class Counter:
+        def __init__(self):
+            self._counter = 0
+
+        def incr(self, n):
+            with FileLock(file_lock[n]):
+                self._counter += 1
+                if fail_flag.exists():
+                    raise Exception()
+
+            if n == 0:
+                return self._counter
+            else:
+                return self.incr.step(n - 1)
+
+        @workflow.virtual_actor.readonly
+        def val(self):
+            return self._counter
+
+        def __getstate__(self):
+            return self._counter
+
+        def __setstate__(self, v):
+            self._counter = v
+
+    locks = [FileLock(f) for f in file_lock]
+    for lock in locks:
+        lock.acquire()
+
+    c = Counter.get_or_create("counter")
+    ray.get(c.ready())
+    final_ret = c.incr.run_async(len(file_lock) - 1)
+    for i in range(0, len(file_lock) - 2):
+        locks[-i - 1].release()
+        val = c.val.run()
+        for _ in range(0, 60):
+            if val == i + 1:
+                break
+            val = c.val.run()
+            time.sleep(1)
+        assert val == i + 1
+
+    fail_flag.touch()
+    locks[1 - len(file_lock)].release()
+    # Fail the pipeline
+    with pytest.raises(Exception):
+        ray.get(final_ret)
+
+    fail_flag.unlink()
+    workflow.resume("counter")
+    # After resume, it'll start form the place where it failed
+    for i in range(len(file_lock) - 1, len(file_lock)):
+        locks[-i - 1].release()
+        val = c.val.run()
+        for _ in range(0, 60):
+            if val == i + 1:
+                break
+            val = c.val.run()
+            time.sleep(1)
+        assert val == i + 1
+
+    assert c.val.run() == 5
+
+    c.write.run(10)
+    assert "\n".join(
+        [str(i) for i in range(10)]) == (tmp_path / "record").read_text()
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular",
+    [{
+        "num_cpus": 4
+        # We need more CPUs, otherwise 'create()' blocks 'get()'
+    }],
+    indirect=True)
+@pytest.mark.skip(
+    reason="Return a list of virtual actor sub method is not supported.")
```

Comment:
@ericl  this one is not supported right now. The reason is that workflow_ref needs to be passed around which is not very easy to do now.
Maybe let's push this kind of support into beta.

Reply:
Hm this seems kind of basic, so what kinds of subworkflows are expected to work?

Reply:
I think it's fine. We can support the following actually:

1. return any kind of normal workflow in virtual actor method
2. inside workflow virtual actor method, call another method and pass a complicated workflow to that method
```
@workflow.virtual_actor
class Actor:
    # what I mean by 1)
    def method(self, param):
         return complicated_workflow.step()

   # what I mean by 2)
   def method2(self):
        start = [complicated_workflow.step(), complicated_workflow_2.step()]
        all = join.step(*start)
       # any kind of workflow is ok
        self.update.step(all)
   
    def update(self, msg):
       # msg is the result for complicated_workflow

    # what can't do
    # self.method.step is not returned directly
    # instead, it returns a workflow, and inside it it's the virtual actor method call (workflow)
    def self_sub_workflow(self):
        all = join.step([self.method.step(), self.method2.step()])
        return all
```

We actually can do a lot of things with just this model:

1. start multiple sessions and update the virtual actor status
2. use virtual actor to trigger any kind of workflow run and they will run one by one

Reply:
In summary:

- return a normal workflow will work
- pass a normal workflow into a virtual actor method will work
- return a virtual actor method inside the virtual actor method will work
- **return a nested workflow with virtual actor method inside it won't work**